In [1]:
# Imported libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix


In [2]:
# Uploaded dataset 
df = pd.read_csv('/datasets/users_behavior.csv')

print(df.head(5))

   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0


In [3]:
# Split features and target data
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra']

In [4]:
# seperating the test data
x_temp, features_test, y_temp, target_test = train_test_split(features, target, test_size=.2, random_state=42)


# training data
features_train, val_test, target_train, val_target = train_test_split(x_temp, y_temp, test_size=.25, random_state=42)




In [5]:
#logistic regression model
model_log = LogisticRegression(random_state=42)
model_log.fit(features_train, target_train)
predictions_log = model_log.predict(val_test)

accuracy_log = accuracy_score(val_target, predictions_log)
precision_log = precision_score(val_target, predictions_log)
recall_log = recall_score(val_target, predictions_log)
f1_log = f1_score(val_target, predictions_log)

print("LOGISTIC REGRESSION PERFORMANCE")
print()
print(f"Accuracy:  {accuracy_log:.1%}")
print(f"Precision: {precision_log:.1%}")
print(f"Recall:    {recall_log:.1%}")
print(f"F1 Score:  {f1_log:.1%}")
print()
print("Confusion Matrix:")
print(confusion_matrix(val_target, predictions_log))


LOGISTIC REGRESSION PERFORMANCE

Accuracy:  72.0%
Precision: 65.0%
Recall:    13.5%
F1 Score:  22.4%

Confusion Matrix:
[[437  14]
 [166  26]]


In [6]:
# Decision tree model
model_tree = DecisionTreeClassifier(random_state=42, max_depth=20, min_samples_split=5, min_samples_leaf=30, max_features='sqrt')
model_tree.fit(features_train, target_train)
predictions_tree = model_tree.predict(val_test)

accuracy_tree = accuracy_score(val_target, predictions_tree)
precision_tree = precision_score(val_target, predictions_tree)
recall_tree = recall_score(val_target, predictions_tree)
f1_tree = f1_score(val_target, predictions_tree)

print("DECISION TREE PERFORMANCE")
print()
print(f"Accuracy:  {accuracy_tree:.1%}")
print(f"Precision: {precision_tree:.1%}")
print(f"Recall:    {recall_tree:.1%}")
print(f"F1 Score:  {f1_tree:.1%}")
print()
print("Confusion Matrix:")
print(confusion_matrix(val_target, predictions_tree))

DECISION TREE PERFORMANCE

Accuracy:  79.5%
Precision: 75.4%
Recall:    46.4%
F1 Score:  57.4%

Confusion Matrix:
[[422  29]
 [103  89]]


In [7]:
# Random forest model
model_for = RandomForestClassifier(random_state=42, n_estimators=300, max_depth=5, min_samples_split=2, min_samples_leaf=5, max_features='sqrt')
model_for.fit(features_train, target_train)
predictions_for = model_for.predict(val_test)

accuracy_for = accuracy_score(val_target, predictions_for)
precision_for = precision_score(val_target, predictions_for)
recall_for = recall_score(val_target, predictions_for)
f1_for = f1_score(val_target, predictions_for)

print("RANDOM FOREST PERFORMANCE")
print()
print(f"Accuracy:  {accuracy_for:.1%}")
print(f"Precision: {precision_for:.1%}")
print(f"Recall:    {recall_for:.1%}")
print(f"F1 Score:  {f1_for:.1%}")
print()
print("Confusion Matrix:")
print(confusion_matrix(val_target, predictions_for))

RANDOM FOREST PERFORMANCE

Accuracy:  80.4%
Precision: 78.9%
Recall:    46.9%
F1 Score:  58.8%

Confusion Matrix:
[[427  24]
 [102  90]]


### Analysis
After running each model and tuning the hyperparameters, the Random Forest Classifier performed the best overall. On the validation dataset, it achieved an accuracy score of 80.4%, meaning it correctly predicted whether a plan was “Ultra” or “Not Ultra” in about 8 out of 10 cases. It also correctly identified Ultra users 78.9% of the time, showing strong performance on the target class. The Decision Tree model was a close second, with an accuracy score of 79.5% and a precision score of 75.4%.

In [8]:
# Best performing model
model = RandomForestClassifier(random_state=42, n_estimators=300, max_depth=5, min_samples_split=2, min_samples_leaf=5, max_features='sqrt')
model.fit(features_train, target_train)
predictions = model.predict(features_test)

accuracy = accuracy_score(target_test, predictions)
precision = precision_score(target_test, predictions)
recall = recall_score(target_test, predictions)
f1 = f1_score(target_test, predictions)

print("BEST PERFORMING MODEL RANDOM FOREST")
print()
print(f"Accuracy:  {accuracy:.1%}")
print(f"Precision: {precision:.1%}")
print(f"Recall:    {recall:.1%}")
print(f"F1 Score:  {f1:.1%}")
print()
print("Confusion Matrix:")
print(confusion_matrix(target_test, predictions))

BEST PERFORMING MODEL RANDOM FOREST

Accuracy:  80.9%
Precision: 78.8%
Recall:    47.3%
F1 Score:  59.1%

Confusion Matrix:
[[431  24]
 [ 99  89]]


In [9]:
# Feature impotance sanity check
feature_importance = model.feature_importances_
feature_names = features.columns

importance_df = pd.DataFrame({
    'Feature': feature_names,
    'Importance': feature_importance,
    'Importance_Percent': feature_importance * 100
}).sort_values('Importance', ascending=False)

print("=== FEATURE IMPORTANCE ANALYSIS ===")
print()
print("features ranked by importance for prediction:")
print(importance_df.round(4))
print()
print("Most Important features:")
for i in range(3):
    feature = importance_df.iloc[i]['Feature']
    percent = importance_df.iloc[i]['Importance_Percent']
    print(f"{i+1}. {feature}: {percent:.1f}% importance")

=== FEATURE IMPORTANCE ANALYSIS ===

features ranked by importance for prediction:
    Feature  Importance  Importance_Percent
3   mb_used      0.3363             33.6329
1   minutes      0.2635             26.3477
0     calls      0.2373             23.7304
2  messages      0.1629             16.2890

Most Important features:
1. mb_used: 33.6% importance
2. minutes: 26.3% importance
3. calls: 23.7% importance


### Analysis
After running the best performing model on the test dataset, the Random Forest Classifier achieved an accuracy score of 80.9% and a precision score of 78.8%. The accuracy score shows that the model correctly predicted whether a user was on the “Ultra” or “Not Ultra” plan in about 8 out of 10 cases. The precision score shows that when the model predicted a user was on the Ultra plan, that prediction was correct 78.8% of the time. Overall, the model performed slightly better on the test dataset than it did on the validation dataset, showing that the Random Forest Classifier generalized well to new data. I then performed a sanity check using feature importance to determine whether the features the model relied on most for its predictions were relevant in real-world scenarios.